File: python/test/exchange/test_ohlcv.py
# -*- coding: utf-8 -*-

__all__ = ['test_ohlcv']


def test_ohlcv(exchange, ohlcv, method, symbol):

    assert isinstance(ohlcv, list)
    assert len(ohlcv) >= 6

    assert isinstance(ohlcv[0], int)  # timestamp

    assert ohlcv[0] > 1230940800000  # 03 Jan 2009 - first block
    assert ohlcv[0] < 2147483648000  # 19 Jan 2038 - int32 overflows

    assert (ohlcv[1] is None) or isinstance(ohlcv[1], (int, float))  # open
    assert (ohlcv[2] is None) or isinstance(ohlcv[2], (int, float))  # high
    assert (ohlcv[3] is None) or isinstance(ohlcv[3], (int, float))  # low
    assert (ohlcv[4] is None) or isinstance(ohlcv[4], (int, float))  # close
    assert (ohlcv[5] is None) or isinstance(ohlcv[5], (int, float))  # volume

    skipped_exchanges = [
        'bitmex',
    ]

    if exchange.id not in skipped_exchanges:
        assert (ohlcv[1] is None) or (ohlcv[2] is None) or (ohlcv[1] <= ohlcv[2]), \
            'open (' + str(ohlcv[1]) + ') > high (' + str(ohlcv[2]) + ')'
        assert (ohlcv[3] is None) or (ohlcv[2] is None) or (ohlcv[3] <= ohlcv[2]), \
            'low (' + str(ohlcv[1]) + ') > high (' + str(ohlcv[2]) + ')'
        assert (ohlcv[3] is None) or (ohlcv[4] is None) or (ohlcv[3] <= ohlcv[4]), \
            'low (' + str(ohlcv[1]) + ') > close (' + str(ohlcv[2]) + ')'

File: ts/src/test/Exchange/test.ohlcv.ts

import assert from 'assert';
import testSharedMethods from './test.sharedMethods.js';

function testOHLCV (exchange, method, entry, symbol, now) {
    const format = [
        1638230400000,
        exchange.parseNumber ('0.123'),
        exchange.parseNumber ('0.125'),
        exchange.parseNumber ('0.121'),
        exchange.parseNumber ('0.122'),
        exchange.parseNumber ('123.456'),
    ];
    const emptyNotAllowedFor = [ 0, 1, 2, 3, 4, 5 ];
    testSharedMethods.assertStructure (exchange, method, entry, format, emptyNotAllowedFor);
    testSharedMethods.assertTimestamp (exchange, method, entry, now, 0);
    const logText = testSharedMethods.logTemplate (exchange, method, entry);
    //
    const length = entry.length;
    assert (length >= 6, 'ohlcv array length should be >= 6;' + logText);
    const skippedExchanges = [
        // 'bitmex', // BitMEX API docs: also note the open price is equal to the close price of the previous timeframe bucket.
    ];
    if (!exchange.inArray (exchange.id, skippedExchanges)) {
        assert ((entry[1] === undefined) || (entry[2] === undefined) || (entry[1] <= entry[2]), 'open > high, ' + exchange.safeString (entry, 1, 'undefined') + ' > ' + exchange.safeString (entry, 2, 'undefined')); // open <= high
        assert ((entry[3] === undefined) || (entry[2] === undefined) || (entry[3] <= entry[2]), 'low > high, ' + exchange.safeString (entry, 2, 'undefined') + ' > ' + exchange.safeString (entry, 3, 'undefined')); // low <= high
        assert ((entry[3] === undefined) || (entry[4] === undefined) || (entry[3] <= entry[4]), 'low > close, ' + exchange.safeString (entry, 3, 'undefined') + ' > ' + exchange.safeString (entry, 4, 'undefined')); // low <= close
    }
    assert ((symbol === undefined) || (typeof symbol === 'string'), 'symbol ' + symbol + ' is incorrect' + logText);
}

export default testOHLCV;
